src/libraries/Common/src/System/Reflection/AssemblyNameParser.cs

```diff
@@ -0,0 +1,568 @@
+        {
+            if (attributeValue.Equals("null", StringComparison.OrdinalIgnoreCase) || attributeValue == string.Empty)
+            {
+                result = Array.Empty<byte>();
+                return true;
+            }
+
+            if (isToken && attributeValue.Length != 8 * 2)
+            {
+                return false;
+            }
+
+            byte[] pkt = new byte[attributeValue.Length / 2];
+            int srcIndex = 0;
+            for (int i = 0; i < pkt.Length; i++)
+            {
+                char hi = attributeValue[srcIndex++];
+                char lo = attributeValue[srcIndex++];
+
+                if (!TryParseHexNybble(hi, out byte parsedHi) || !TryParseHexNybble(lo, out byte parsedLo))
+                {
+                    return false;
+                }
+
+                pkt[i] = (byte)((parsedHi << 4) | parsedLo);
+            }
+            result = pkt;
+            return true;
+        }
+
+        private static bool TryParseProcessorArchitecture(string attributeValue, out ProcessorArchitecture result)
+        {
+            result = attributeValue switch
+            {
+                _ when attributeValue.Equals("msil", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.MSIL,
+                _ when attributeValue.Equals("x86", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.X86,
+                _ when attributeValue.Equals("ia64", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.IA64,
+                _ when attributeValue.Equals("amd64", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.Amd64,
+                _ when attributeValue.Equals("arm", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.Arm,
+                _ when attributeValue.Equals("msil", StringComparison.OrdinalIgnoreCase) => ProcessorArchitecture.MSIL,
+                _ => ProcessorArchitecture.None
+            };
+            return result != ProcessorArchitecture.None;
+        }
+
+        private static bool TryParseHexNybble(char c, out byte parsed)
+        {
+            int value = HexConverter.FromChar(c);
+            if (value == 0xFF)
+            {
+                parsed = 0;
+                return false;
+            }
+            parsed = (byte)value;
+            return true;
+        }
+
+        private static bool IsWhiteSpace(char ch)
+        {
+            switch (ch)
+            {
+                case '\n':
+                case '\r':
+                case ' ':
+                case '\t':
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool TryGetNextChar(out char ch)
+        {
+            if (_index < _input.Length)
+            {
+                ch = _input[_index++];
+                if (ch == '\0')
```

Comment:
I'm curious why this case is different from just hitting the end of the input?

Reply:
I've not authored this particular method (I've moved the parser from https://github.com/dotnet/runtime/blob/a2ce8db5fb14df85cef03880a30133be198626e1/src/libraries/System.Private.CoreLib/src/System/Reflection/AssemblyNameParser.cs#L307-L324 and extended with a possibility to not throw on error and apply strict rules). 

My understanding is that we treat `\0` as end of the string, no matter the index.

Reply:
This is one of the compat warts of the CoreLib parser. We should do this in the CoreLib flavor of the parser only. I see you do that in a few other places already: https://github.com/dotnet/runtime/pull/100094/files#diff-07c55a8a24f36769b3f5daa10e97dac14684ebd28197aa5e5d71a9152c86a218R192

If we cared enough about the parsers behaving the same without these warts, I think it would be fine to take a breaking change for cases like this one.